### ISSUE TITLE
Cast not applied for Coroutine and generic class

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pyright#7936]
**Describe the bug**
Applying `cast` in the case below doesn't seem to be applied to the variable. I would expect the second reveal_type to just be `A[Coroutine[Any, Any, _R@bad]`.
```py
from typing import Any, Coroutine, cast

class A[_T]:
    def __init__(self, x: _T) -> None: ...

def bad[_R](
    a1: A[Coroutine[Any, Any, _R] | _R],
) -> None:
    reveal_type(a1)
    a1 = cast(A[Coroutine[Any, Any, _R]], a1)
    reveal_type(a1)
```
```py
  test:9:17 - information: Type of "a1" is "A[Coroutine[Any, Any, _R@bad] | _R@bad]"
  test:11:17 - information: Type of "a1" is "A[Coroutine[Any, Any, _R@bad] | _R@bad]"
```

In contrast, this works fine:
```py
def good[_R](
    a2: A[list[_R] | _R],
    a3: Coroutine[Any, Any, _R] | _R,
) -> None:
    reveal_type(a2)
    a2 = cast(A[list[_R]], a2)
    reveal_type(a2)

    reveal_type(a3)
    a3 = cast(Coroutine[Any, Any, _R], a3)
    reveal_type(a3)
```
```py
  test.py:18:17 - information: Type of "a2" is "A[list[_R@good] | _R@good]"
  test.py:20:17 - information: Type of "a2" is "A[list[_R@good]]"
  test.py:22:17 - information: Type of "a3" is "Coroutine[Any, Any, _R@good] | _R@good"
  test.py:24:17 - information: Type of "a3" is "Coroutine[Any, Any, _R@good]"
```

**VS Code extension or command-line**
pyright 1.1.363

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The typing spec doesn't (currently) say anything about how type checkers are supposed to apply type narrowing upon assignment to a variable with a declared type. Mypy and pyright differ somewhat in how and when they apply such narrowing. Some of these differences are [documented](https://microsoft.github.io/pyright/#/mypy-comparison?id=assignment-based-type-narrowing).

The rules for type narrowing upon assignment get especially murky when the declared type and/or the assigned type include gradual typing constructs like `Any` or `...` (in a Callable). In cases where the declared type includes `Any`, pyright always attempts to preserve `Any`. If the assigned type includes `Any` but the declared type does not (as in the example above), pyright uses a more complex algorithm to determine whether to apply type narrowing.

I think pyright's behavior here is defensible, consistent with the typing spec, and internally consistent. I could be convinced to change it in the future if we decide to standardize the behavior. Until then, I'm going to leave it as is because I don't think it would be beneficial to current pyright users who would simply see it as churn.
